src/jdk.incubator.jextract/share/classes/jdk/internal/jextract/impl/DeclarationImpl.java

```diff
@@ -57,25 +62,41 @@ public Position pos() {
         return pos;
     }
 
+    @Override
+    public Optional<List<ConstantDesc>> getAttribute(String name) {
+        return Optional.ofNullable(attributes.get(name));
+    }
+
+    @Override
+    public Set<String> attributeNames() { return Collections.unmodifiableSet(attributes.keySet()); }
+
     public static class VariableImpl extends DeclarationImpl implements Declaration.Variable {
 
         final Variable.Kind kind;
         final Type type;
         final Optional<MemoryLayout> layout;
 
+        private VariableImpl(Type type, Optional<MemoryLayout> layout, Variable.Kind kind, String name, Position pos, Map<String, List<ConstantDesc>> attrs) {
```

Comment:
Rather than changing these factories to add attribute variants, consider either using `withAttribute` and drop the changes, or add an internal method in DeclarationImp which clones a declaration and adds annotations (even using a single Map).